### PR TITLE
`printf`: Fix extra padding space and wrong position of `-` when padding

### LINKED ISSUE #6519: `printf`: extra spaces if padding number is not next to %
Cases: 
printf "% -3.0f" 1
 1 %
cargo run printf "% -3.0f" 1
 1  %(extra spaces here)
cargo run printf '%-3.0f' 1`
1  %

printf "% 3.0d" 1
 1 %
cargo run printf "% 3.0d" 1
 1  %(extra spaces here)

However, if the number is negative, the space will be filled with `-`, then there's no extra padding:
cargo run printf "% -3.0f" -1
-1 %

related issue: #2774

### PR DESCRIPTION
**Pull request by Kev1n8**
should fix #6519 and #2774. 
added some padding tests and fix the incorrect behaviors of padding

### CODE REVIEW

File: src/uucore/src/lib/features/format/num_format.rs [hunk 1]
```diff
@@ -78,20 +78,36 @@ impl Formatter for SignedInt {
...
+        } else {
+            s = s[1..].to_string();
+            String::from("-")
+        };
 
+        let remaining_width = if self.width >= flag.len() {
+            self.width - flag.len()
+        } else {
+            0
+        };
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:102]: please add some comments explaining what it is doing

File: src/uucore/src/lib/features/format/num_format.rs [hunk 2]
```diff
@@ -272,10 +280,30 @@ impl Formatter for Float {
             format_float_non_finite(x, self.case)
         };
 
+        let flag = if x >= 0. {
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:283]: seems duplicated code with above, could you please have a look to create a function ?
thanks

File: src/uucore/src/lib/features/format/num_format.rs (hunk 1 above)
Comment (Kev1n8) [reply src/uucore/src/lib/features/format/num_format.rs]: Sure. Just added some comments.

File: src/uucore/src/lib/features/format/num_format.rs (hunk 2 above)
Comment (Kev1n8) [reply src/uucore/src/lib/features/format/num_format.rs]: I created 2 functions and considered combining them using traits, but that seems unnecessary. Let me know if there's any other way to improve.

File: src/uucore/src/lib/features/format/num_format.rs [hunk 3]
```diff
@@ -326,6 +340,30 @@ impl Formatter for Float {
     }
 }
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:342]: what about something like 
```
fn get_flag<T: PartialOrd + Default>(sign: PositiveSign, x: T) -> String {
    if x >= T::default() {
        match sign {
            PositiveSign::None => String::new(),
            PositiveSign::Plus => String::from("+"),
            PositiveSign::Space => String::from(" "),
        }
    } else {
        String::from("-")
    }
}
```

Comment (Kev1n8) [reply src/uucore/src/lib/features/format/num_format.rs:342]: Nice one! I didn't know about that. Now updated.

**Comment by Kev1n8** [comment]
Just added a case check, should fix the remaining issue in #2774. Before this, there was a incorrect output as following:
Screenshot 2024-07-06 at 15 22 30

**Comment by sylvestre** [comment]
> Just added a case check, should fix the remaining issue in #2774. Before this, there was a incorrect output as following: Screenshot 2024-07-06 at 15 22 30

could you please replace the screenshot by text? it is terrible for accessibility and search. thanks

File: src/uucore/src/lib/features/format/num_format.rs [hunk 4]
```diff
@@ -78,20 +79,33 @@ impl Formatter for SignedInt {
...
+        // by storing remaining_width indicating the actual width needed.
+        // Using min() because self.width could be 0, 0usize - 1usize should be avoided
+        let remaining_width = self.width - min(self.width, flag.len());
         match self.alignment {
-            NumberAlignment::Left => write!(writer, "{s:<width$}", width = self.width),
-            NumberAlignment::RightSpace => write!(writer, "{s:>width$}", width = self.width),
-            NumberAlignment::RightZero => write!(writer, "{s:0>width$}", width = self.width),
+            NumberAlignment::Left => write!(writer, "{flag}{s:<width$}", width = remaining_width),
+            NumberAlignment::RightSpace => {
+                let is_sign_flag = flag.starts_with("-") || flag.starts_with("+"); // When flag is in ['-', '+']
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:97]:
```suggestion
                let is_sign_flag = flag.starts_with('-') || flag.starts_with('+'); // When flag is in ['-', '+']
```

File: src/uucore/src/lib/features/format/num_format.rs [hunk 5]
```diff
@@ -272,10 +278,32 @@ impl Formatter for Float {
...
+        } else {
+            0
+        };
         match self.alignment {
-            NumberAlignment::Left => write!(writer, "{s:<width$}", width = self.width),
-            NumberAlignment::RightSpace => write!(writer, "{s:>width$}", width = self.width),
-            NumberAlignment::RightZero => write!(writer, "{s:0>width$}", width = self.width),
+            NumberAlignment::Left => write!(writer, "{flag}{s:<width$}", width = remaining_width),
+            NumberAlignment::RightSpace => {
+                let is_sign_flag = flag.starts_with("-") || flag.starts_with("+"); // When flag is in ['-', '+']
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:295]:
```suggestion
                let is_sign_flag = flag.starts_with('-') || flag.starts_with('+'); // When flag is in ['-', '+']
```

File: src/uucore/src/lib/features/format/num_format.rs [hunk 6]
```diff
@@ -272,10 +278,32 @@ impl Formatter for Float {
...
+        let remaining_width = if self.width >= flag.len() {
+            self.width - flag.len()
+        } else {
+            0
+        };
         match self.alignment {
-            NumberAlignment::Left => write!(writer, "{s:<width$}", width = self.width),
-            NumberAlignment::RightSpace => write!(writer, "{s:>width$}", width = self.width),
-            NumberAlignment::RightZero => write!(writer, "{s:0>width$}", width = self.width),
+            NumberAlignment::Left => write!(writer, "{flag}{s:<width$}", width = remaining_width),
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:293]: you have duplicated code here, could you please dedup it ? thanks

**Comment by Kev1n8** [comment]
> could you please replace the screenshot by text? it is terrible for accessibility and search. thanks
Sorry for my carelessness. The incorrect output text is as following:
/coreutils printf "%+5.1f" 2
+ 2.0%

File: src/uucore/src/lib/features/format/num_format.rs (hunk 6 above)
Comment (Kev1n8) [reply src/uucore/src/lib/features/format/num_format.rs]: done by using a `write_output` function.

File: src/uucore/src/lib/features/format/num_format.rs [hunk 7]
```diff
@@ -326,6 +315,18 @@ impl Formatter for Float {
     }
 }
 
+fn get_flag<T: PartialOrd + Default>(sign: PositiveSign, x: &T) -> String {
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:318]: get_flag is a bit too generic 
maybe call it get_sign_indicator ?

Comment (Kev1n8) [reply src/uucore/src/lib/features/format/num_format.rs:318]: Indeed. Changed it now. thanks for the advice.

**Comment by Kev1n8** [comment]
sorry, that was an accident.

**Pull request closed without merging**